File: code/python/spiralOrder.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        if matrix == []:
            return []
        left_bound = 0
        right_bound = len(matrix[0])
        down_bound = len(matrix)
        up_bound = 0
        res = []
        i = j = 0
        while True:
            #right            
            while j < right_bound:
                res.append(matrix[i][j])
                j = j + 1
            up_bound += 1
            if i == down_bound-1:
                return res
            
            #down
            j = j - 1
            i = i + 1
            while i < down_bound:
                res.append(matrix[i][j])
                i = i + 1
            right_bound -= 1
            if j == left_bound:
                return res
            
            #left
            i = i - 1
            j = j - 1
            while j >= left_bound:
                res.append(matrix[i][j])
                j -= 1
            down_bound -= 1
            j = j + 1
            if i == up_bound:
                return res
            
            #up
            i = i - 1
            while i >= up_bound:
                res.append(matrix[i][j])
                i = i - 1
            left_bound += 1
            if j == right_bound-1:
                return res
            i = i + 1
            j = j + 1

File: code/java/spiralOrder.java
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        List<Integer> res = new ArrayList<Integer>();
        if(matrix.length == 0)
            return res;
        int i=0,j=0;
        int leftBound=0, rightBound=matrix[0].length - 1, upBound=0, downBound = matrix.length - 1;
        while(true){
            //go right
            while(j <= rightBound)
                res.add(matrix[i][j++]);
            upBound++;
            if(i == downBound)
                return res;
            j--;
            i++;
            //go down
            while(i <= downBound)
                res.add(matrix[i++][j]);
            rightBound--;
            if(j == leftBound)
                return res;
            i--;
            j--;
            //go left
            while(j >= leftBound)
                res.add(matrix[i][j--]);
            downBound--;
            if(i == upBound)
                return res;
            j++;
            i--;
            //go up
            while(i >= upBound)
                res.add(matrix[i--][j]);
            leftBound++;
            if(j==rightBound)
                return res;
            i++;
            j++;
            
        }
    }
}
